### ISSUE TITLE
poor peer-peer performance in Tr2.94, Tr3, Tr4

### ISSUE THREAD

**Issue by reardonia** [transmission/transmission#6986]
I know there have been several transfer-perf related issues filed in the past, but I'm filing this one with a more controlled look at transfer-perf in hopes that if we make progress on this one form of the problem (simple two-peer swarm), the more complex bandwidth issues will follow.

Tr4-Tr4 transfer rates are worse than Tr4-Libtorrent, and signficantly worse the Libtorrent-Libtorrent, ~~or Tr3-Tr3.~~

The Test: two clients running on the same machine Linux machine talking thru loopback interface. No other torrents loaded, clients pointed at each other via a local tracker.  Transfer 12GB file, read from RAM disk, write to RAM disk.

Results:
1) Tr4 <-> Tr4: 16MB/s, sometimes dropping to 8MB/s. Very weird, but seems to be very binary 16 or 8.  Over a 1Gbit link, throughput is generally 8MB/s sometimes dropping to 4MB/s.  
2) Tr4(up) <-> LT2(down): 16MB/s.
3) Tr4(down) <-> LT2(up): 58-63MB/s.  
4) LT2 <-> LT2: 235-240MB/s
5) Tr3.00 <-> Tr3.00: 8MB/s or 4MB/s.
5b) Tr2.94 <-> Tr2.94: same as Tr3-Tr3.
5c) Tr2.52 <->  Tr2.52: same
5d) Tr2.03 <-> Tr2.03: 4MB/s. Notable as this was before uTP support.
6) Tr4.0.6 <-> Tr4.0.6: 8MB/s with drops to 4MB/s.  Very binary 4 or 8.

Some notes:
1) "Tr4" means mainline 4.1 branch, testing done from ec5296c8d
2) Release builds (with symbols).
3) Tr3-Tr3 is preliminary, need to reverify.
4) Did this on three very different Linux machines: a) recent Cascade Lake-W, b) older Broadwell, c) Ryzen 7. memory from 8GB to 64GB.  
5) Cache at 4MB-64MB, makes very little difference.
6) Libtorrent client was qBittorrent 4.5.2 with Libtorrent 2.0.8
7) uTP was disabled in all clients.  The test uses TCP, Encryption is enabled and negotiated. Chose this config because it represents most-common-case.
8) Speeds are very consistent after ~5 seconds of rampup.
9) tested with Debian distribution builds just to make sure this wasn't an artifact of my personal builds.  Same results.
10) tested Tr3.00 on Windows 10 using QT build.  Same issue, maxed out at 16MB/s, though slightly more variable. This tells us it's not a weird Linux/Unix socket issue.
11) Tested Tr2.52, same.

There seems to be concensus within the posted Issues here on github that Tr4 is substantially slowed the Tr3, and this is clearly the case. I wanted to provide a relatively clinical look at perf in a controlled network environment.

Tr4 seems to have acute problems with upload bandwidth, though it also has problems with download. I thought that by reducing the problem to a simple two-peer swarm we could start tackling this more aggressively and measureably. Curious to hear thoughts from @tearfur among others.

UPDATE 2024-07-11
+ added Tr4.06 results.  
+ retested with UTP just as quick sanity check.  same peak results, though more consistent. So 8MBs-4MBs oscillation becomes strictly 8MBs.
UPDATE 2024-07-12
+ Tr3 is pretty bad too.  I'm not sure how I got higher results the first time I tried, so I'll assume I logged incorrectly. This is an important result, because it's a direct rebuttal to a lot of the "Tr4 perf is much worse than Tr3" posts I saw.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

the Issue occurs in 2.94, 3.00, and 4.0.6 release, as well as Tr4.1 mainline.

**Comment by tearfur** [comment]
So Tr4-Tr4 speeds are often powers of 2...?

Also I'm curious about the difference between uTP and TCP connections.

**Comment by Pentaphon** [comment]
Something got "lost in translation" during the refactor to C++ and is causing Tr4 to be locked into these wildly varying and slow speeds. 

>Also I'm curious about the difference between uTP and TCP connections.

I find that disabling uTP speeds up Tr4 but the speed still goes up and down while completing the transfer instead of always going up until the transfer is complete like it should.

**Comment by tearfur** [comment]
Let me get this straight before the discussion gets too messy:

This thread is strictly limited to discussions about **mainline** Tr4, and tests between 2 clients.

**Comment by reardonia** [comment]
> So Tr4-Tr4 speeds are often powers of 2...?

Yes.  On all machines tested speeds were 4MB/s, 8MB/s, or 16MB/s. Could be spurious, but notable. The other variable was linespeed, which was either 1Gbit or loopback.

> Also I'm curious about the difference between uTP and TCP connections.

Should have noted that this is TCP only.  I wanted to keep the test as simple and reproducible as possible.  I'll add further notes to original description. The point I was making in #6975 / #6971 made it hard to test this, but also exposed an announcer bug I'll file separately.

**Comment by reardonia** [comment]
> t's fine if you mention previous Tr4 release for comparison with the nightlies. But if you are going to discuss about detailed behaviour, please keep _**nightly**_ Tr4 versions as the main subject.

Yes, agree.  Tried to make all of this clear in the Issue description.  Happy to change the title if you have suggestion.  But especially interested to know if you think it's worth focusing on transfer-perf from perspective of this 2-client, controlled setup.

Now that you bring it up, seems like I should add 4.0.6 release as well.

**Comment by reardonia** [comment]
Updated with 4.0.6 results.

**Comment by tearfur** [comment]
Hmm, there's definitely something weird going on, probably with the upload code.

I tried to repeat your test in https://github.com/tearfur/transmission/tree/tmp, and the speed is ***hard capped*** at 35.55 MB/s depsite having no speed limit set. The speed number was rock solid.

![image](https://github.com/user-attachments/assets/139fec09-9efe-42b1-816c-32eaaf37e578)

**Comment by reardonia** [comment]
Updated with 3.00 results.

**Comment by reardonia** [comment]
Updatd with 2.94 results.

**Comment by reardonia** [comment]
As this is likely a problem deep in `tr_bandwidth`, I did a variety of experients on `phase_one`, varying `Increment` from 3000U up to 8000000U.  Increasing `Increment` to 65536U or more makes the transfer rate more reliable (eg it stays very consistently at 16MB/s and does not periodically vary).

Curious why the Increment value was decided for uTP but we use it for TCP/classic connections?

**Comment by reardonia** [comment]
Got Transmission 2.03 from 2010 (!!) working on Debian and tested.  This is from long before UTP support was added.  It ALSO gets stuck exactly at 4MB/s.

**Comment by Pentaphon** [comment]
> Got Transmission 2.03 from 2010 (!!) working on Debian and tested. This is from long before UTP support was added. It ALSO gets stuck exactly at 4MB/s.

It has always been my suspicion that the libtransmission code hasn't been optimized from the ground up since 2.0 because of how long the code had been on the old git that github only recently replaced, and now that torrent backend development has generally slowed down and has shifted to libtorrent-rasterbar (which serves as the backend for several popular clients), the chances of a ground-up optimization for libtransmission has become even less likely these days. I sincerely hope that this eye-opening testing by reardonia sparks enough interest to have libtransmission get the modern update it deserves. 

>Curious why the Increment value was decided for uTP but we use it for TCP/classic connections?

I think nobody bothered to do the testing to find an optimal uTP value so they just used the value for TCP connections, which goes to show how long this project has gone without proper testing and optimization.

**Comment by reardonia** [comment]
> > Got Transmission 2.03 from 2010 (!!) working on Debian and tested. This is from long before UTP support was added. It ALSO gets stuck exactly at 4MB/s.
> 
> It has always been my suspicion that the libtransmission code hasn't been optimized from the ground up since 2.0 ...

Transmission 1.22 (the oldest I can get working, and it's truly bizarre to use, involves hacking with unix sockets) is much worse, perf-wise

**Comment by tearfur** [comment]
Just brainstorming, haven't had time to look into this more deeply yet.

I think this is a more likely bottleneck than the bandwidth code:

https://github.com/transmission/transmission/blob/a76a07ae992eb7a31435b7087c8ecfba50c1c33c/libtransmission/peer-msgs.cc#L1790-L1806

> Increasing `Increment` to 65536U or more makes the transfer rate more reliable (eg it stays very consistently at 16MB/s and does not periodically vary).

Well, this shows the bottleneck isn't (just) the bandwidth code.



> > Curious why the Increment value was decided for uTP but we use it for TCP/classic connections?
> 
> I think nobody bothered to do the testing to find an optimal uTP value so they just used the value for TCP connections, which goes to show how long this project has gone without proper testing and optimization.

No no, that value isn't even used for uTP right now, but the comment in the code says that the value was decided with uTP performance in mind. As for why, I don't know (yet).

**Comment by Pentaphon** [comment]
> and the speed is _**hard capped**_ at 33.55 MB/s depsite having no speed limit set. The speed number was rock solid.

I never noticed this before but I just did some high speed tests and I couldn't get over this number with either 4.0.6 or the latest nightly. This is pretty huge. I'd imagine that every user with good upload will want this fixed.

**Comment by tearfur** [comment]
Just an update. It's pretty clear to me that the reason for the 16MB/s cap for Tr<->Tr transfers is because our client `reqq` (BEP-10) value is hard-coded at 512.

https://github.com/transmission/transmission/blob/3e9f5f614a875aae987235d3836d0ca555a80d80/libtransmission/peer-msgs.cc#L181

The rest is simple maths:

```
max upload speed = 512 * 16KiB / 0.5s = 16KiB/s
```

[^1]: BT piece data are sent in 16KiB blocks.
[^2]: For each peer, Transmission sends a new batch of outgoing block request every 0.5 seconds.

**Comment by Pentaphon** [comment]
> Just an update. It's pretty clear to me that the reason for the 16MB/s cap for Tr<->Tr transfers is because our client `reqq` (BEP-10) value is hard-coded at 512.

According to https://www.bittorrent.org/beps/bep_0010.html

>`reqq` = An integer, the number of outstanding request messages this client supports without dropping any. The default in libtorrent is 250.

Is it time to test different values or go with what libtorrent does?

**Comment by tearfur** [comment]
Raising the value will solve a big part of the problem.

And there's always more to it 😉

**Comment by Pentaphon** [comment]
> Raising the value will solve a big part of the problem.

I see from libtorrent they have 

>if (reqq > 0) max_out_request_queue(reqq);

https://github.com/arvidn/libtorrent/blob/d3eef89002dfd4499a7fb7f42d5adbc5aa180556/src/bt_peer_connection.cpp#L2067

**Comment by tearfur** [comment]
That doesn't mean much, it's just a fail-safe against an invalid value.

I expect to have a PR up soon that raises this value to 2000 (matching libtorrent) and allow users to configure this value as needed. But that will take a while since I want to test it with other changes I have in the works.

**Comment by Rukario** [comment]
> allow users to configure this value as needed.

It would be weird to discover an issue and then we get the new option in next release asking us to control it. Sure we can make it automatic?

**Comment by Roardom** [comment]
> Libtorrent clients (qbt, deluge) are able to easily saturate a 1gigabit connection, so >100MB/s. If upload speed corresponds so directly to REQQ, seems like you'd be stuck at 64MB/s peak vs current situation. No?

For what it's worth, qBittorrent can generally saturate a 10 gigabit connection, and the max speed in a single swarm I've seen was Deluge 1.2 reaching 32 gigabit/s upload.

**Comment by Rukario** [comment]
> That doesn't mean much, it's just a fail-safe against an invalid value.
> 
> I expect to have a PR up soon that raises this value to 2000 (matching libtorrent) and allow users to configure this value as needed. But that will take a while since I want to test it with other changes I have in the works.

Just curious, why is 2000 such a definite number. What kind of benefit will we have by controlling it?

**Comment by tearfur** [comment]
> If upload speed corresponds so directly to REQQ, seems like you'd be stuck at 64MB/s peak vs current situation. No?

Yes, so that's why other changes are needed to further improve it. Right now, I am exploring the these 2 ideas:

1. Have Transmission send out block requests as soon as it gets a chance. e.g. After getting a piece/reject/unchoke message.
2. Reply with a piece message as soon as it receives a request message.

**Comment by tearfur** [comment]
Here is a test from the changes I currently have. You can see that it easily reaches 120+ MiB/s now.

This seems to be hammering Transmission's session thread so hard that I get no updates at all from the WebUI during the download. So the GIF recording is taken from qB's UI.

- Tr4(down) <-> LT2.0.10.0(up)
- 4GB torrent
- Tr4 built with RelWithDebInfo
- TCP connection

![tmp](https://github.com/user-attachments/assets/b20e09e8-5a38-4cc4-99fc-d9320955f45a)

**Comment by Pentaphon** [comment]
> * 4GB torrent

Very interested to see exactly what changes you came up with. 1 question though. Why use a 4GB file instead of say, a 16GB file to observe the speed for a longer period of time?

**Comment by tearfur** [comment]
Here's the latest progress for Tr4's improving download speed.

In my previous test, the wishlist was way too slow to support downloading from multiple peers, it slows down to an abysmal speed. So this time I put in some hard work to optimise it.

- Tr4(down) built with RelWithDebInfo
- TCP connection
- Prefers encryption

### Ubuntu torrent

Slightly better than the current `main` on my machine.

- 1G Internet

![ubuntu](https://github.com/user-attachments/assets/ff02712b-ce4d-4a44-9685-1047169e7959)

![flame-ubuntu](https://github.com/user-attachments/assets/f31e8dc8-54f5-40ec-b536-b9af071fe4d9)

### Local machine

Speeds further improved as the wishlist is now faster. Top upload speed recorded on qBit was 190 MiB/s.

The Transmission WebUI still freezes, but this probably can't be helped until we have async IO.

- LT2.0.10.0(up)
- Single connection
- Both clients on the same Linux machine, seeding from and downloading to the same tmpfs.
- Incomplete dir disabled

![local](https://github.com/user-attachments/assets/ce93ab69-62e6-4b0f-8063-2ce7cb82b6ba)

![flame-local](https://github.com/user-attachments/assets/6ab9628d-da5c-4693-ab71-56b163d6d19d)

**Comment by tearfur** [comment]
Now for upload speed, I managed to get Tr4 to upload at 200+ MB/s.

- Tr4(down) <-> Tr4(up)
- RelWithDebInfo
- TCP connection
- Prefers encryption
- Single connection
- Both clients on the same Linux machine, seeding from and downloading to the same tmpfs
- Incomplete dir disabled

![tr](https://github.com/user-attachments/assets/31d9a6f7-de55-4f33-b10b-6a7283c43718)

![flame-tr](https://github.com/user-attachments/assets/305a3687-7c8f-4ba4-9f68-a39b21cbe33b)

**Comment by Pentaphon** [comment]
>The Transmission WebUI still freezes, but this probably can't be helped until we have async IO.

Does the WebUI freeze during real-world torrenting or just in your local network? I'm wondering if these new optimizations have to be postponed in order to not freeze real-world user WebUIs until async I/O is implemented, which is probably not happening anytime soon.

**Comment by tearfur** [comment]
@Pentaphon Just in my local network.

In general, it'll probably tend to freeze less the larger the round trip time is, and freeze more the more peers there are.

**Comment by Pentaphon** [comment]
This doesn't matter anymore but earlier today, I was using  4.0.6 connected to another sole 4.0.6 peer in real world usage and the speed keeps bouncing between 4 and 8MB/s, back and forth over and over. Amazing how I never noticed this before. I recorded it on my webui. 

https://github.com/user-attachments/assets/ff8e3bda-165f-4bf4-945b-c9afa9579bce

**Comment by reardonia** [comment]
> @Pentaphon Just in my local network.
> 
> In general, it'll probably tend to freeze less the larger the round trip time is, and freeze more the more peers there are.

I haven't dug into, but if I remember the old Transmission event loop, RPC and peer events are treated equally, ie the code just relies on libevent to put everything into one big queue. in Tr4, to avoid these RPC (and thus UI) lockups, would it be possible to run two separate queues, and when servicing the peer queue always do some kind of lookahead into the RPC queue?  in other words, always prioritize RPC traffic?

**Comment by tearfur** [comment]
> but if I remember the old Transmission event loop, RPC and peer events are treated equally, ie the code just relies on libevent to put all inbound packets into one big queue.

IIRC this is still true in Tr4.



> would it be possible to run two separate queues, and when servicing the peer queue always do some kind of lookahead into the RPC queue? in other words, always prioritize RPC traffic?

I know that libevent can set priorities for events, but Tr never used it. That could be a direction to investigate.

**Comment by tearfur** [comment]
Went through another round of optimisation and bug fixes, now Tr is able to saturate my 1G Internet connection with ease.

If you look closely to the flame graph, you'll notice that the big block of `Wishlist::next` that was previously there is now so small that you don't see the text anymore.

* Tr4(down) built with RelWithDebInfo
* TCP connection
* Prefers encryption

![ubuntu](https://github.com/user-attachments/assets/d1b1b833-6ea3-476f-856c-cab04afef357)

![flame](https://github.com/user-attachments/assets/7e5d2039-de79-4212-a476-1e91bfa38ac9)

**Comment by tearfur** [comment]
The test in my previous comment was supposed to be for a version of 7027+7026, but turns out that there was a bug that caused Tr to behave like some middle ground of 7027 and 7026.

In proper 7027+7026, your download speed should be limited only by your disk IO, bandwidth, and single-core CPU speed. When I fixed the bug, the big `Wishlist::next` block in the flame graph came back, and my test machine became CPU bottlenecked again, dropping the download speed back down.

In the middle ground, the CPU load during download is quite a bit lower than proper 7027+7026. However, it also has the hard single-peer speed limit that was the original subject of this issue. On my test machine it just so happened that the lower CPU load gave room for a peer-saturated swarm to fill more of my bandwidth, but I imagine that the download speeds in this middle ground can only get so fast even if I have lightning-fast disk IO and download.

Still, 7027 is a major performance improvement to active request tracking and the wishlist. So for now, until anyone has an idea for 7026, I'm going to submit the middle ground in 7027, and leave 7026 as draft.

**Comment by reardonia** [comment]
> The test in my previous comment was supposed to be for a version of 7027+7026, but turns out that there was a bug that caused Tr to behave like some middle ground of 7027 and 7026.
> 
> In proper 7027+7026, your download speed should be limited only by your disk IO, bandwidth, and single-core CPU speed. When I fixed the bug, the big `Wishlist::next` block in the flame graph came back, and my test machine became CPU bottlenecked again, dropping the download speed back down.
> 
> In the middle ground, the CPU load during download is quite a bit lower than proper 7027+7026. However, it also has the hard single-peer speed limit that was the original subject of this issue. On my test machine it just so happened that the lower CPU load gave room for a peer-saturated swarm to fill more of my bandwidth, but I imagine that the download speeds in this middle ground can only get so fast even if I have lightning-fast disk IO and download.
> 
> Still, 7027 is a major performance improvement to active request tracking and the wishlist. So for now, until anyone has an idea for a better version of 7026, I'm going to submit the middle ground in 7027, and leave 7026 as draft.

Is the "middle ground" submitted in your #7027 update now?  (on holiday so apologies for slow responses thru end of month)

**Comment by tearfur** [comment]
Yes.

**Comment by nagisa** [comment]
Out of curiosity, have you investigated how much of an overhead popcount compiling down to `__popcountdi2` instead of `popcnt` is? And how much that affects the transfer speeds? Should be a relatively easy test (with `-mcpu=native` or some such.) If its a significant improvement, it might be worthwhile to build multiple versions of affected functions and dispatch to them dynamically, based on runtime availability of the instructions (popcnt is ~universally available nowadays.)

**Comment by tearfur** [comment]
You can check `libtransmission/tr-popcount.h`. It sounds a bit like what you are asking for.
